data/tutorials/language/0lg_00_basic_datatypes.md

```diff
@@ -0,0 +1,808 @@
+- : t2 = T2 (4, false)
+```
+
+The constructor `T1` has two parameters of type `int` and `bool`, whilst the constructor `T2` has a single parameter of type `int * bool`.
+
+#### Recursive Variants
+
+A variant definition referring to itself is recursive. A constructor may wrap data from the type being defined.
+
+This is the case for the following definition, which can be used to store JSON values.
+```ocaml
+# type json =
+  | Null
+  | Bool of bool
+  | Int of int
+  | Float of float
+  | String of string
+  | Array of json list
+  | Object of (string * json) list;;
+type json =
+    Null
+  | Bool of bool
+  | Int of int
+  | Float of float
+  | String of string
+  | Array of json list
+  | Object of (string * json) list
+```
+
+Both constructors `Array` and `Object` contain values of type `json`.
+
+Functions defined using pattern matching on recursive variants are often recursive too. This function checks if a name is present in a whole JSON tree:
+```ocaml
+# let rec has_field name = function
+  | Array u ->
+      List.fold_left (fun b obj -> b || has_field name obj) false u
+  | Object u ->
+      List.fold_left (fun b (key, obj) -> b || key = name || has_field name obj) false u
+  | _ -> false;;
+val has_field : string -> json -> bool = <fun>
+```
+
+Here, the last pattern uses the symbol `_`, which catches everything. It returns `false` on all data that is neither `Array` nor `Object`.
+
+### Polymorphic Data Types
+
+#### Revisiting Predefined Types
+
+The predefined type `option` is a variant type with two constructors: `Some` and `None`. It can contain values of any type, such as `Some 42` or `Some "hola"`. In that sense, `option` is polymorphic. Here is how it is defined in the standard library:
+
+```ocaml
+# #show option;;
+type 'a option = None | Some of 'a
+```
+
+The predefined type `list` is polymorphic in the same sense. It is a variant with two constructors and can hold data of any type. Here is how it is defined in the standard library:
+```ocaml
+# #show list;;
+type 'a list = [] | (::) of 'a * 'a list
+```
+
+The only magic here is turning constructors into symbols, which we don't cover in this tutorial. The types `bool` and `unit` also are regular variants, with the same magic:
+```ocaml
+# #show unit;;
+type unit = ()
+
+# #show bool;;
+type bool = false | true
+```
+
+Implicitly, product types also behave as variant types. For instance, pairs can be seen as inhabitants of this type:
+```ocaml
+# type ('a, 'b) pair = Pair of 'a * 'b;;
+type ('a, 'b) pair = Pair of 'a * 'b
+```
+
+`(int, bool) pair` would be written `int * bool`, and `Pair (42, true)` would be written `(42, true)`. From the developer's perspective, everything happens as if such a type were declared for every possible product shape. This is what allows pattern matching on products.
+
+Even integers and floats can be seen as enumerated-like variant types, with many constructors and funky syntactic sugar, which allows pattern matching on those types.
```

Comment:
Nice, you didn't killed “funky”.

Is this the right place for an easter egg?
Maybe this one: https://youtu.be/O0_H3F84Yjk

Reply:
HELL YES! 😂 🕺